### PR TITLE
Select prim without data source must not crash.

### PR DESCRIPTION
**Pull request by ppt-adsk**

### CODE REVIEW

File: lib/flowViewport/sceneIndex/fvpSelectionSceneIndex.cpp [hunk 1]
```diff
@@ -76,7 +76,7 @@ class _PrimSource : public HdContainerDataSource
 
     TfTokenVector GetNames() override
     {
-        TfTokenVector names = _inputSource->GetNames();
+        TfTokenVector names = _inputSource ? _inputSource->GetNames() : TfTokenVector();
```
Comment (ppt-adsk) [inline lib/flowViewport/sceneIndex/fvpSelectionSceneIndex.cpp:79]: Allow input data source to be null.

File: lib/flowViewport/sceneIndex/fvpSelectionSceneIndex.cpp [hunk 2]
```diff
@@ -132,7 +132,12 @@ SelectionSceneIndex::GetPrim(const SdfPath &primPath) const
         .Msg("SelectionSceneIndex::GetPrim() called.\n");
 
     HdSceneIndexPrim result = GetInputSceneIndex()->GetPrim(primPath);
-    if (!result.dataSource) {
+    // An empty data source can be taken to mean an invalid prim, but it can
+    // also be an ancestor in a path to an actual prim (see
+    // https://forum.aousd.org/t/representation-of-invalid-hdsceneindexprim/833/2
+    // ).  Such an ancestor prim can be selected, even if typeless, so add
+    // a selection data source in such a case.
+    if (!result.dataSource && !_selection->IsFullySelected(primPath)) {
```
Comment (ppt-adsk) [inline lib/flowViewport/sceneIndex/fvpSelectionSceneIndex.cpp:140]: The bug fix.  The crash was caused by a failing assertion in the wireframe selection highlighting scene index, triggered by the fact that switching selection to a prim without a data source did not add a selection data source to that prim.

File: test/lib/mayaUsd/render/mayaToHydra/testSelectPrimWithoutDataSource.py
Comment (ppt-adsk) [inline test/lib/mayaUsd/render/mayaToHydra/testSelectPrimWithoutDataSource.py:1]: Without the fix, this test crashes.  With the fix, it passes.

File: test/testSamples/testSelectPrimWithoutDataSource/cube.usda
Comment (ppt-adsk) [inline test/testSamples/testSelectPrimWithoutDataSource/cube.usda:1]: Trivial model with cards draw mode support.  This creates a parent without a data source, which is how the crash gets triggered.

File: test/testSamples/testSelectPrimWithoutDataSource/cube.usda [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    defaultPrim = "cube"
+)
+
+def Xform "cube" (
+    prepend apiSchemas = ["GeomModelAPI"]
+    kind = "component"
+)
+{
+    uniform bool model:applyDrawMode = 1
+    asset model:cardTextureXNeg = @./red20x20.png@
```
Comment (ppt-adsk) [inline test/testSamples/testSelectPrimWithoutDataSource/cube.usda:12]: All card axes have the same texture in our test, which is unimportant.  A real model would have different textures.

File: test/testSamples/testSelectPrimWithoutDataSource/root.usda [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+        double3 xformOp:rotateXYZ = (-90, 0, 0)
+        double3 xformOp:translate = (-5, 0, 0)
+        uniform token[] xformOpOrder = ["xformOp:rotateXYZ", "xformOp:translate"]
+    }
+    def Xform "refAssetB" (
+        references = @./cube.usda@
+        kind = "group"
+    )
+    {
+        uniform token model:drawMode = "cards"
```
Comment (ppt-adsk) [inline test/testSamples/testSelectPrimWithoutDataSource/root.usda:24]: This causes refAssetB to be drawn in card mode, which triggers the bug.

**Review by debloip-adsk — APPROVED** [review]

**Pull request merged**